File: tools/openclaw-hermes-router.js
#!/usr/bin/env node
/**
 * OpenClaw -> Hermes HTTP router
 *
 * Local-only HTTP server. Not exposed externally.
 * Endpoint: POST /invoke
 * Body: { message: string, conversationId?: string }
 * Response: { ok, reply, stderr, returnCode }
 */
const http = require('node:http');
const { spawn } = require('node:child_process');

const PORT = parseInt(process.env.OPENCLAW_HERMES_ROUTER_PORT || '18790', 10);
const HOST = process.env.OPENCLAW_HERMES_ROUTER_HOST || '127.0.0.1';
const HERMES_TIMEOUT_MS = parseInt(process.env.OPENCLAW_HERMES_TIMEOUT_MS || '120000', 10);

async function invokeHermes(message, retries = 2) {
  const args = ['chat', '-q', message];
  let lastError = null;
  let attempt = 0;

  while (attempt < retries) {
    attempt += 1;
    try {
      const proc = spawn('hermes', args, {
        stdio: ['ignore', 'pipe', 'pipe'],
        env: { ...process.env, HOME: process.env.HOME || 'C:\\Users\\HP' },
      });

      let stdout = '';
      let stderr = '';
      let settled = false;

      const timer = setTimeout(() => {
        if (!settled) {
          settled = true;
          proc.kill('SIGTERM');
          setTimeout(() => {
            if (!proc.killed) proc.kill('SIGKILL');
          }, 5000);
        }
      }, HERMES_TIMEOUT_MS);

      proc.stdout.on('data', (chunk) => { stdout += chunk.toString(); });
      proc.stderr.on('data', (chunk) => { stderr += chunk.toString(); });

      const exitCode = await new Promise((resolve) => {
        proc.on('close', (code) => {
          clearTimeout(timer);
          resolve(code);
        });
        proc.on('error', (err) => {
          clearTimeout(timer);
          resolve(-1);
        });
      });

      const trimmed = stdout.trim();
      if (exitCode === 0 && trimmed) {
        return { ok: true, reply: trimmed, stderr: stderr.trim(), returnCode: exitCode };
      }

      lastError = { ok: false, reply: '', stderr: stderr.trim() || `Empty output from hermes (exit ${exitCode})`, returnCode: exitCode };
    } catch (e) {
      lastError = { ok: false, reply: '', stderr: String(e), returnCode: -1 };
    }
  }

  return lastError || { ok: false, reply: '', stderr: 'hermes unavailable', returnCode: -1 };
}

async function handleRequest(req, res) {
  if (req.method !== 'POST' || req.url !== '/invoke') {
    res.writeHead(404, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ error: 'Not found' }));
    return;
  }

  let body = '';
  req.on('data', (chunk) => { body += chunk.toString(); });

  req.on('end', async () => {
    let payload = {};
    try {
      payload = body ? JSON.parse(body) : {};
    } catch (e) {
      res.writeHead(400, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: `Invalid JSON: ${e.message}` }));
      return;
    }

    const message = payload.message || payload.query || '';
    if (!message) {
      res.writeHead(400, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ error: 'Missing message' }));
      return;
    }

    const result = await invokeHermes(message, 2);
    const status = result.ok ? 200 : 502;
    res.writeHead(status, { 'Content-Type': 'application/json' });
    res.end(JSON.stringify(result));
  });
}

const server = http.createServer(handleRequest);
server.listen(PORT, HOST, () => {
  process.stdout.write(`OPENCLAW_HERMES_ROUTER_LISTEN ${HOST}:${PORT}\n`);
});

process.on('SIGINT', () => {
  process.stdout.write('OPENCLAW_HERMES_ROUTER_SHUTDOWN\n');
  server.close(() => process.exit(0));
});

File: tools/openclaw_hermes_router.py
#!/usr/bin/env python3
"""
HTTP adapter for OpenClaw → Hermes router.

This adapter exposes router behavior in testable form without importing Node
files from Python. Integration tests should call this via subprocess or use
the Node test suite.
"""

import json
import os
import subprocess
import sys
import urllib.request
from pathlib import Path

BASE = Path(__file__).resolve().parent.parent
ROUTER_SCRIPT = BASE / "tools" / "openclaw-hermes-router.js"


def _invoke_local(message: str, timeout: int = 120) -> dict:
    if not ROUTER_SCRIPT.exists():
        return {"ok": False, "reply": "", "stderr": f"missing script: {ROUTER_SCRIPT}", "returnCode": -1}
    try:
        proc = subprocess.run(
            ["node", str(ROUTER_SCRIPT)],
            input=json.dumps({"message": message}),
            capture_output=True,
            text=True,
            timeout=timeout,
        )
        stdout = (proc.stdout or "").strip()
        stderr = (proc.stderr or "").strip()
        if proc.returncode == 0 and stdout:
            try:
                parsed = json.loads(stdout)
                return parsed
            except Exception:
                return {"ok": True, "reply": stdout, "stderr": stderr, "returnCode": proc.returncode}
        return {"ok": False, "reply": "", "stderr": stderr or f"Empty output from router (exit {proc.returncode})", "returnCode": proc.returncode}
    except subprocess.TimeoutExpired:
        return {"ok": False, "reply": "", "stderr": "router invocation timed out", "returnCode": -1}
    except Exception as e:
        return {"ok": False, "reply": "", "stderr": str(e), "returnCode": -1}


def _post(path: str, payload: dict, timeout: int = 30) -> dict:
    router_host = os.getenv("OPENCLAW_HERMES_ROUTER_HOST", "127.0.0.1")
    router_port = int(os.getenv("OPENCLAW_HERMES_ROUTER_PORT", "18790"))
    url = f"http://{router_host}:{router_port}{path}"
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(url, data=data, headers={"Content-Type": "application/json"}, method="POST")
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            try:
                return json.loads(resp.read().decode("utf-8"))
            except Exception:
                return {"ok": True, "status": int(resp.status)}
    except urllib.error.HTTPError as e:
        try:
            body = e.read().decode("utf-8")
        except Exception:
            body = ""
        return {"ok": False, "status": int(e.code), "error": body}
    except Exception as e:
        return {"ok": False, "status": 0, "error": str(e)}


def invoke_hermes(message: str, retries: int = 2) -> dict:
    last = None
    for _ in range(max(1, retries)):
        last = _invoke_local(message)
        if last.get("ok"):
            return last
    return last or {"ok": False, "reply": "", "stderr": "router unavailable", "returnCode": -1}


def send_telegram_message(text: str, to: str = None) -> dict:
    return _post("/invoke", {"message": text, "query": text})


def main():
    try:
        data = json.load(sys.stdin)
    except Exception as e:
        print(json.dumps({"error": f"Invalid input JSON: {e}"}))
        sys.exit(1)

    message = data.get("message") or data.get("query") or ""
    if not message:
        print(json.dumps({"error": "Missing message"}))
        sys.exit(1)

    result = invoke_hermes(message, retries=int(data.get("retries", 2)))
    print(json.dumps(result, indent=2))
    sys.exit(0 if result.get("ok") else 2)


if __name__ == "__main__":
    main()
